### PR TITLE
trunner: add sparcv8leon3-generic-qemu target support

### PR DESCRIPTION
**Pull request by maska989**
JIRA: CI-474

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [X] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [X] Tested by hand on: (sparcv8leon3-generic-qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing linter checks and tests passed.
- [X] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [X] This PR needs additional PRs to work
https://github.com/phoenix-rtos/phoenix-rtos-project/pull/1168
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: mem/test.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ test:
       ignore: True #issue #1010 https://github.com/phoenix-rtos/phoenix-rtos-project/issues/1010
       targets:
         # mprotect: is noop on NOMMU architecture
-        exclude: [armv7m7-imxrt106x-evk, armv7m7-imxrt117x-evk, armv7m4-stm32l4x6-nucleo]
+        exclude: [armv7m7-imxrt106x-evk, armv7m7-imxrt117x-evk, armv7m4-stm32l4x6-nucleo,sparcv8leon3-generic-qemu]
```
Comment (damianloew) [inline mem/test.yaml:9]: We want to add a target as experimental for now as we only run psh tests, please see
https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/366#discussion_r1721838618

**Review by damianloew — CHANGES_REQUESTED** [review]
Please enable this target to CI in a separate commit.

File: trunner/target/emulated.py [hunk 2]
```diff
@@ -80,9 +85,22 @@ def from_context(cls, _: TestContext):
         return cls()
 
 
+class SPARCV8Leon3GenericQemuTarget(QemuTarget):
+    name = "sparcv8leon3-generic-qemu"
+    rootfs = False
+
```
Comment (mateusz-bloch) [inline trunner/target/emulated.py:91]:
```suggestion
    rootfs = False
    experimental = True
```

And then include `sparcv8leon3-generic-qemu` to psh tests

**Review by mateusz-bloch — COMMENTED** [review]
When building with tests option

```
LD  test_mmap               
/src/phoenix-rtos-tests/devices/grlib-multi/test-grlib-multi.c:28:2: error: #error "Unsupported target"
   28 | #error "Unsupported target"
      |  ^~~~~
/src/phoenix-rtos-tests/devices/grlib-multi/test-grlib-multi.c:31:10: fatal error: grlib-multi.h: No such file or directory
   31 | #include <grlib-multi.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
make: *** [../phoenix-rtos-build/Makefile.common:211: /src/_build/sparcv8leon3-generic-qemu/phoenix-rtos-tests/devices/grlib-multi/test-grlib-multi.o] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by mateusz-bloch** [comment]
> When building with tests option
> 
> ```
> LD  test_mmap               
> /src/phoenix-rtos-tests/devices/grlib-multi/test-grlib-multi.c:28:2: error: #error "Unsupported target"
>    28 | #error "Unsupported target"
>       |  ^~~~~
> /src/phoenix-rtos-tests/devices/grlib-multi/test-grlib-multi.c:31:10: fatal error: grlib-multi.h: No such file or directory
>    31 | #include <grlib-multi.h>
>       |          ^~~~~~~~~~~~~~~
> compilation terminated.
> make: *** [../phoenix-rtos-build/Makefile.common:211: /src/_build/sparcv8leon3-generic-qemu/phoenix-rtos-tests/devices/grlib-multi/test-grlib-multi.o] Error 1
> make: *** Waiting for unfinished jobs....
> ```

You could add this to the Makefile.sparcvleon3, or come up with better solution

```
ifneq ($(TARGET), sparcv8leon3-generic-qemu)
DEFAULT_COMPONENTS += test-grlib-multi
endif
```

File: psh/test.yaml [hunk 3]
```diff
@@ -13,7 +13,7 @@ test:
         - name: pshlogin
           harness: test-pshlogin.py
           targets:
```
Comment (mateusz-bloch) [inline psh/test.yaml:15]: On the beginning of yaml should be smth like:
```yaml
test:
    targets:
      #TODO: remove when sparcv8leon3-generic-qemu target will stop being experimental
      include: [sparcv8leon3-generic-qemu]
```

File: trunner/target/emulated.py [hunk 4]
```diff
@@ -22,7 +27,9 @@ def __init__(self, script: str):
         super().__init__()
         self.script = script
         # TODO Make sure that script path exists
-        self.dut = QemuDut(f"{self.project_dir}/scripts/{self.script}", encoding="utf-8")
+        self.dut = QemuDut(
+            f"{self.project_dir}/scripts/{self.script}", encoding="utf-8"
+        )
```
Comment (mateusz-bloch) [inline trunner/target/emulated.py:32]: Nit: For future: changes about  formatting of untouched parts of  file should not be part of the same pr

File: psh/test.yaml [hunk 5]
```diff
@@ -1,4 +1,7 @@
 test:
+    targets:
+        value: [sparcv8leon3-generic-qemu]
```
Comment (mateusz-bloch) [inline psh/test.yaml:3]: You should include target, not set sparcv8leon3-generic-qemu as only target, also comment should be place why it is done.
```suggestion
        include: [sparcv8leon3-generic-qemu]
```

File: Makefile.sparcv8leon3 [hunk 6]
```diff
@@ -1 +1,3 @@
+ifneq ($(TARGET), sparcv8leon3-generic-qemu)
```
[the remark this answers is not in the document]
Comment (mateusz-bloch) [reply Makefile.sparcv8leon3:1]: Could also do this in a separate commit and be connected to this issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/1182

[the remark this answers is not in the document]
Comment (damianloew) [reply Makefile.sparcv8leon3:1]: Yeah, it should be done in the separate commit for sure

[the remark this answers is not in the document]
Comment (damianloew) [reply Makefile.sparcv8leon3:1]: @maska989 please create a separate commit for excluding this test from build

**Review by damianloew — CHANGES_REQUESTED** [review]

**Review by damianloew — COMMENTED** [review]
Some notes about commit descriptions.

As it's described in our `Golden Rules`

 - describe WHY You're doing the change in the commit message body

So you don't want to repeat what you have added in the commit message body (called also commit description). In your case the reason of adding a new target is rather obvious - we just want to support all of our available targets.

About the second commit `sparcv8leon3: prevent from building test-grlib-multi on qemu target` - the commit title shows perfectly what is going on in this change, you probably want to add information WHY do we need to exclude it, so a good commit description would be:

```
This test is not intended to build on this particular target.
```

**Review by damianloew — APPROVED** [review]

**Pull request merged**